pkg/runctl/controller.go

```diff
@@ -265,13 +265,23 @@ func (c *Controller) syncHandler(key string) error {
 		return nil
 	}
 
+	// Init state when undefined
+	if pipelineRun.GetStatus().State == api.StateUndefined {
+		err = pipelineRun.InitState()
+		if err != nil {
+			return err
+		}
+	}
+
 	// Check if object has deletion timestamp
 	// If not, try to add finalizer if missing
 	if pipelineRun.HasDeletionTimestamp() {
 		runManager := c.createRunManager(pipelineRun)
 		err = runManager.Cleanup(pipelineRun)
+
 		if err == nil {
-			err = pipelineRun.DeleteFinalizerIfExists()
+			pipelineRun.UpdateResult(api.ResultDeleted)
```

Comment:
Where is the finalizer deleted now?

Reply:
The `finish` function removes the finalizer.

Reply:
OK, in `c.finish`.